File: mcp/src/velum_mcp/analysis/ai_analyzer.py
"""AI-powered task failure analysis.

Uses the MCP host's LLM context — when called from Claude, Claude itself
analyses the task output and provides a diagnosis. No external API key needed.
"""

from __future__ import annotations

from typing import Any

from velum_mcp.client import get_client


async def analyze_task_failure(
    project_id: int,
    task_id: int,
    last_n_lines: int = 100,
) -> dict[str, Any]:
    """Analyse a failed task's output and return a structured diagnosis.

    This tool retrieves the task output and structures it so the AI assistant
    (Claude) can provide a diagnosis within its context window — no external
    API key needed.

    Args:
        project_id: Project containing the task.
        task_id: The failed task to analyse.
        last_n_lines: How many lines of output to include (default 100).

    Returns:
        A dict with: task_metadata, output_tail, analysis_prompt.
        The calling AI should use analysis_prompt to generate the diagnosis.
    """
    client = get_client()

    # Fetch task metadata
    task = await client.get(f"/projects/{project_id}/tasks/{task_id}")
    status = task.get("status", "unknown")

    if status not in ("error", "stopped"):
        return {
            "error": f"Task {task_id} has status '{status}' — analysis is for failed tasks only.",
            "task": task,
        }

    # Fetch output
    try:
        raw = await client.get(f"/projects/{project_id}/tasks/{task_id}/raw_output")
        if isinstance(raw, str):
            lines = raw.splitlines()
        elif isinstance(raw, list):
            lines = [str(r) for r in raw]
        else:
            lines = []
    except Exception:  # noqa: BLE001
        lines = []

    output_tail = "\n".join(lines[-last_n_lines:]) if lines else "(no output captured)"

    return {
        "task_id": task_id,
        "project_id": project_id,
        "template_id": task.get("template_id"),
        "status": status,
        "started": task.get("start"),
        "finished": task.get("end"),
        "build_task": task.get("build_task"),
        "output_lines_total": len(lines),
        "output_tail": output_tail,
        "analysis_prompt": (
            f"The following Velum task (ID {task_id}) failed with status '{status}'.\n"
            f"Template ID: {task.get('template_id')}\n"
            f"Started: {task.get('start')}, Finished: {task.get('end')}\n\n"
            f"Last {min(last_n_lines, len(lines))} lines of output:\n"
            f"```\n{output_tail}\n```\n\n"
            "Please:\n"
            "1. Identify the root cause of the failure\n"
            "2. List the most likely reasons (ranked by probability)\n"
            "3. Provide specific remediation steps\n"
            "4. Suggest any preventive measures for the future"
        ),
    }


async def bulk_analyze_failures(
    project_id: int,
    limit: int = 5,
) -> list[dict[str, Any]]:
    """Fetch and structure the N most recent failed tasks for bulk analysis.

    Args:
        project_id: Project to scan.
        limit: Number of recent failures to include (default 5).

    Returns:
        List of task summaries ready for AI analysis.
    """
    client = get_client()
    tasks: list[dict[str, Any]] = await client.get(f"/projects/{project_id}/tasks") or []
    failed = [t for t in tasks if t.get("status") == "error"]
    # Most recent first
    failed = sorted(failed, key=lambda t: t.get("id", 0), reverse=True)[:limit]

    results = []
    for task in failed:
        try:
            summary = await analyze_task_failure(project_id, task["id"], last_n_lines=50)
            results.append(summary)
        except Exception as e:  # noqa: BLE001
            results.append({"task_id": task["id"], "error": str(e)})

    return results

File: mcp/src/tools/ai_analyzer.rs
/// AI-powered task failure analysis.
///
/// Retrieves task output and structures it as a rich prompt so the
/// calling AI assistant (Claude) can diagnose failures in its own context —
/// no external API key or second LLM call required.
use crate::client::VelumClient;
use crate::protocol::{prop_int, prop_int_opt, Tool, ToolResult};
use serde_json::{json, Value};
use anyhow::Result;

pub fn definitions() -> Vec<Tool> {
    vec![
        Tool {
            name: "analyze_task_failure".into(),
            description: "Retrieve output from a failed task and structure it for AI diagnosis. \
                Returns the task metadata and console output so you (Claude) can identify \
                the root cause and suggest fixes — no external API key needed.".into(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "project_id": prop_int("Project ID"),
                    "task_id": prop_int("The failed task ID to analyse"),
                    "last_n_lines": prop_int_opt("Lines of output to include (default 100)")
                },
                "required": ["project_id","task_id"]
            }),
        },
        Tool {
            name: "bulk_analyze_failures".into(),
            description: "Fetch the N most recent failed tasks and structure them for bulk AI analysis. \
                Returns metadata + output snippets for each failure.".into(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "project_id": prop_int("Project ID"),
                    "limit": prop_int_opt("Number of recent failures to include (default 5)")
                },
                "required": ["project_id"]
            }),
        },
    ]
}

pub async fn call(name: &str, args: &Value, client: &VelumClient) -> Result<ToolResult> {
    let pid = args["project_id"].as_i64().unwrap_or_default();
    match name {
        "analyze_task_failure" => {
            let tid = args["task_id"].as_i64().unwrap_or_default();
            let n = args.get("last_n_lines").and_then(Value::as_u64).unwrap_or(100) as usize;
            let result = fetch_failure_analysis(client, pid, tid, n).await?;
            Ok(ToolResult::json(&result))
        }
        "bulk_analyze_failures" => {
            let limit = args.get("limit").and_then(Value::as_u64).unwrap_or(5) as usize;
            let all = client.get(&format!("/projects/{pid}/tasks")).await?;
            let failed: Vec<Value> = all.as_array().map(|a| {
                let mut f: Vec<Value> = a.iter()
                    .filter(|t| t["status"].as_str() == Some("error"))
                    .cloned()
                    .collect();
                f.sort_by_key(|t| -(t["id"].as_i64().unwrap_or(0)));
                f.truncate(limit);
                f
            }).unwrap_or_default();

            let mut results = Vec::new();
            for task in &failed {
                if let Some(tid) = task["id"].as_i64() {
                    match fetch_failure_analysis(client, pid, tid, 50).await {
                        Ok(r) => results.push(r),
                        Err(e) => results.push(json!({"task_id": tid, "error": e.to_string()})),
                    }
                }
            }
            Ok(ToolResult::json(&json!(results)))
        }
        _ => Ok(ToolResult::error(format!("Unknown tool: {name}"))),
    }
}

async fn fetch_failure_analysis(client: &VelumClient, pid: i64, tid: i64, n: usize) -> Result<Value> {
    let task = client.get(&format!("/projects/{pid}/tasks/{tid}")).await?;
    let status = task["status"].as_str().unwrap_or("unknown");

    if status != "error" && status != "stopped" {
        return Ok(json!({
            "error": format!("Task {tid} has status '{status}' — analysis is for failed tasks only."),
            "task": task
        }));
    }

    let raw = client.get(&format!("/projects/{pid}/tasks/{tid}/raw_output")).await.unwrap_or(Value::Null);
    let lines: Vec<String> = match &raw {
        Value::String(s) => s.lines().map(String::from).collect(),
        Value::Array(arr) => arr.iter().map(|r| r.as_str().unwrap_or("").to_string()).collect(),
        _ => vec![],
    };

    let total_lines = lines.len();
    let tail: Vec<&String> = lines.iter().rev().take(n).collect::<Vec<_>>().into_iter().rev().collect();
    let output_tail = tail.iter().map(|s| s.as_str()).collect::<Vec<_>>().join("\n");
    let displayed = tail.len();

    Ok(json!({
        "task_id": tid,
        "project_id": pid,
        "template_id": task.get("template_id"),
        "status": status,
        "started": task.get("start"),
        "finished": task.get("end"),
        "total_output_lines": total_lines,
        "output_tail": output_tail,
        "analysis_guidance": format!(
            "Task #{tid} failed (status: {status}). Template ID: {}. \
             Started: {}, Finished: {}.\n\n\
             Last {displayed}/{total_lines} lines of console output:\n\
             ```\n{output_tail}\n```\n\n\
             Please diagnose:\n\
             1. Root cause of the failure\n\
             2. Most likely reasons (ranked by probability)\n\
             3. Specific remediation steps\n\
             4. Preventive measures for the future",
            task.get("template_id").and_then(Value::as_i64).unwrap_or(0),
            task.get("start").and_then(Value::as_str).unwrap_or("unknown"),
            task.get("end").and_then(Value::as_str).unwrap_or("unknown"),
        )
    }))
}
